### PR TITLE
KAFKA-8206: Allow client to rebootstrap

### PR DESCRIPTION
**Pull request by ivanyu**
This commit implements KIP-899: Allow producer and consumer clients to rebootstrap. It introduces the new setting `metadata.recovery.strategy`, applicable to all the types of clients.

The rebootstrapping is tested on the integration level.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:1]: I added a number of constructors here instead of altering the existing because they're public and I assume a part of the client library contract.

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java]: This section is very confusing for me, and I think this should be reworded.

Maybe it can flow like this: "Metadata recovery is triggered when all last-known brokers appear unavailable simultaneously. Brokers appear unavailable when disconnected and no current retry attempt is in-progress."

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (gharris1727) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:1]: Stuff in this package is not included in the public API: https://kafka.apache.org/37/javadoc/index.html only the stuff in admin/producer/consumer (and not in sub-packages) is truly public.

**Review by gharris1727 — COMMENTED** [review]
Thanks so much for the fix @ivanyu!

I noticed there isn't an AdminClientReboostrapTest, is that intentional? I wouldn't personally require one because the change is in NetworkClient and not the admin client specifically, but just wanted to check with you.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:1]: Thank you. I removed extra constructors in this case

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java]: This makes sense. Thank you, updated the doc

**Comment by gharris1727** [comment]
Hi @ivanyu Could you resolve the merge conflicts?

**Comment by ivanyu** [comment]
Yep, doing this + finishing with the admin client

**Comment by ivanyu** [comment]
> I noticed there isn't an AdminClientReboostrapTest, is that intentional? I wouldn't personally require one because the change is in NetworkClient and not the admin client specifically, but just wanted to check with you.

Yes, sorry. It was moved in/out of scope a couple of times during discussion, so I forgot to add the implementation and test for it.

Also, the conflict resolved

**Comment by ivanyu** [comment]
I see there are some failures in the rebootstrap tests on CI. I need to deal with them

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:309]: Is there a reason we are re-bootstrapping from addresses that were resolved when the client was created? This seems different from the example in the KIP where the addresses that the bootstrap resolved to had changed.

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: In other enums like ClientDnsLookup, we throw IllegalArgumentException, we could do the same here.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Couldn't `node == null` mean that all the connections have max.in.flight requests and hence we need to wait?

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminMetadataManager.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminMetadataManager.java:302]: Same question as with NetworkClient, don't we want to re-resolve bootstrap?

File: core/src/main/scala/kafka/raft/RaftManager.scala
Comment (rajinisivaram) [inline core/src/main/scala/kafka/raft/RaftManager.scala:316]: Rather than change all these callers, couldn't we retain the original constructor for these?

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (rajinisivaram) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Which clients in the broker do we want to rebootstrap? Don't remember this change from the KIP.

File: core/src/test/scala/integration/kafka/api/RebootstrapTest.scala
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/RebootstrapTest.scala:39]: We have moved away from using fixed ports in tests since they can result in spurious test failures. But not sure what the alternative would be for these.

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: nit: this is unused

**Review by rajinisivaram — COMMENTED** [review]
@ivanyu Thanks for the PR, looks good overall. Left some comments and questions.

File: core/src/test/scala/integration/kafka/api/RebootstrapTest.scala
Comment (gharris1727) [reply core/src/test/scala/integration/kafka/api/RebootstrapTest.scala:39]: If Kafka had a development-only option to reuse addresses, we could reserve ports ahead-of-time, and have the broker startup just rebind the address.

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: nit: this can NPE if the input is null. This currently can't happen, but it would be nice to get the same exception type for all invalid inputs.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:309]: This really depends on the `client.dns.lookup`. E.g. if you use the bootstrap address `kafka.myorg.com:12345`, the bootstrap node will get `kafka.myorg.com` as the host, despite the name resolution happens before that.

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: Some others like `TimestampType` throw `NoSuchElementException`, but indeed `IllegalArgumentException` is probably better. Fixed

File: core/src/main/scala/kafka/raft/RaftManager.scala
Comment (ivanyu) [reply core/src/main/scala/kafka/raft/RaftManager.scala:316]: I originally did this. It basically doubled the number of constructors and looked hairy. After [this discussion](https://github.com/apache/kafka/pull/13277#discussion_r1607019116) reverted to the current approach

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: Yes, removed

File: clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/MetadataRecoveryStrategy.java]: Yep, it handle this now

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (ivanyu) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: You're right, it's not needed here. Removed

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminMetadataManager.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminMetadataManager.java:302]: It uses the same `ClientUtils.parseAndValidateAddresses` function inside, so the same logic applies.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Ugh, that's correct. I'm sorry, I missed this case...

I need to think how to pass this information to avoid rebootstrapping. Potentially introduce a variant of `leastLoadedNode` method, which will be called from two places where rebootstrapping is expected. It's result will have a flag indicating if any node is connected. The rest of the code base will use the current version.

The necessity to implement this for the admin client, which is different from other clients, makes it a bit ugly :)

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:309]: Yes, this seems fine because we are only using host string from the InetSocketAddress to build the bootstrap cluster.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: I have an idea how to avoid rebootstrapping in this case. The idea is to pass some information along with the `Node` out of `leastLoadedNode`. See this commit https://github.com/apache/kafka/pull/13277/commits/3a303b7de650a6d6a94d5652a476ca9a98610380. Does it make sense? It changes the interface a little, but the overall change turned out to be smaller than I expected.

File: server/src/main/java/org/apache/kafka/server/config/ServerConfigs.java
Comment (rajinisivaram) [inline server/src/main/java/org/apache/kafka/server/config/ServerConfigs.java]: These are no longer required?

File: clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java]: We are setting this to true only for ready connections, should we rename the method to indicate that?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Could we add a method to `LeastLoadedNode` that checks `leastLoadedNode.node() == null && !leastLoadedNode.isAtLeastOneConnected()` since we use that here and in Admin client? It will make it more obvious in LeastLoadedNode that the boolean is only used when node is null. We could then remove `isAtLeastOneConnected` method perhaps.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Can we add one unit test for the various combinations to verify the returned values for LeastLoadedNode?

**Review by rajinisivaram — COMMENTED** [review]
@ivanyu Thanks for the updates, LGTM. Left a few minor comments. Would you be able to address those today/tomorrow so that we can try and get this merged for 3.8.0?

**Comment by rajinisivaram** [comment]
@ivanyu Also need to fix unit tests that are failing because of the change from Node to LeastLoadedNode and resolve merge conflicts.

File: server/src/main/java/org/apache/kafka/server/config/ServerConfigs.java
Comment (ivanyu) [reply server/src/main/java/org/apache/kafka/server/config/ServerConfigs.java]: Right, removed

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Great suggestion, done

File: clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java]: Done this, but https://github.com/apache/kafka/pull/13277#discussion_r1633579132 deprecates this anyway

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Right, I added some unit testing

**Comment by ivanyu** [comment]
Comments addressed and conflicts resolved

**Review by rajinisivaram — APPROVED** [review]
@ivanyu Thanks for the PR, LGTM. Will merge to trunk if build is good. Also following up on 3.8.0 thread to check if it can be merged to the 3.8 branch.

**Comment by rajinisivaram** [comment]
@ivanyu There are more test failures that need fixing before this PR can be merged.

File: clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java]: This can just use `node` field.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Is there a chance that there may be connections to old nodes? I think we want to rebootstrap only if every node is in backoff state. We don't close nodes in the Admin client equivalent below, which is confusing, so it may be better to remove this unless it is required. Checking `hasNodeAvailableOrConnectionReady` adds to the confusion since it suggests there may be nodes in some non-ready state that need to be closed. Should we change `leastLoadedNode()` method to return `hasConnectedOrConnecting` nodes instead of `atLeastOneConnectionReady` to make this more obvious?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Should we swap this around to `if (metadataRecoveryStrategy == MetadataRecoveryStrategy.REBOOTSTRAP && !leastLoadedNode.hasNodeAvailableOrConnectionReady())` here and in the admin client, so that we only use the new method if rebootstrap is configured?

**Review by rajinisivaram — COMMENTED** [review]
@ivanyu Left a few more minor comments as well, can you take a look when you are fixing the tests?

File: clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/LeastLoadedNode.java]: Yeah, done

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: As it's only comparisons inside the method, performance-wise it doesn't matter much. However, it's more logical. Done this.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
Comment (ivanyu) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Oh that's right: if `!leastLoadedNode.hasNodeAvailableOrConnectionReady()`, there shouldn't be open connections. I removed this closing.

**Comment by ivanyu** [comment]
Some tests are failing on one pipeline, but they seem unrelated

**Comment by rajinisivaram** [comment]
@ivanyu Thanks for the updates, test failures not related, merging to trunk and 3.8.

**Pull request merged**
